verify-logging/main.go

```diff
@@ -0,0 +1,150 @@
+// Package main contains the logic for using Cloud Monitoring to determine whether requests have been receiving 5xx errors.
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"os"
+	"time"
+
+	monitoring "cloud.google.com/go/monitoring/apiv3/v2"
+	"cloud.google.com/go/monitoring/apiv3/v2/monitoringpb"
+	"github.com/golang/protobuf/ptypes/timestamp"
+	"google.golang.org/api/iterator"
+)
+
+var (
+	// Variable to hold the flag's values.
+	project            string
+	metricFilter       string
+	maxErrorPercentage float64
+	triggerDuration    time.Duration
+	timeToMonitor      time.Duration
+	samplingPeriod     time.Duration
+	samplingWindow     time.Duration
+)
+
+func init() {
+	// Initializing of the flag and print out the values for visibility.
+	flag.StringVar(&project, "project", "", "The ID of the project that has the metrics")
+	flag.StringVar(&metricFilter, "metric-filter", "", "A [monitoring filter](https://cloud.google.com/monitoring/api/v3/filters) that specifies which time series should be returned")
+	flag.Float64Var(&maxErrorPercentage, "max-error-percentage", 10, "The maximum allowable percentage of 5xx response_code_class per sampling")
+	flag.DurationVar(&triggerDuration, "trigger-duration", 5*time.Minute, "The time required to observe the error condition for verify to fail")
+	flag.DurationVar(&timeToMonitor, "time-to-monitor", 20*time.Minute, "The time to monitor for response failures before the verification is marked successful")
+	flag.DurationVar(&samplingPeriod, "sampling-period", time.Minute, "The time to wait in between each sampling")
+	flag.DurationVar(&samplingWindow, "sampling-window", 5*time.Minute, "The window of time that specifies the dataset for each sampling")
+	flag.Parse()
+
+	fmt.Println("---")
+	fmt.Println("Verification configured as follows:")
+	fmt.Printf("Project: %q\n", project)
+	fmt.Printf("Metric Filter: %q\n", metricFilter)
+	fmt.Printf("Max Error Percentage: %v\n", maxErrorPercentage)
+	fmt.Printf("Trigger Duration: %v\n", triggerDuration)
+	fmt.Printf("Time To Monitor: %v\n", timeToMonitor)
+	fmt.Printf("Sampling Period: %v\n", samplingPeriod)
+	fmt.Printf("Sampling Window: %v\n", samplingWindow)
+	fmt.Println("---")
+}
+
+const (
+	responseLabelName    = "response_code_class"
+	responseCodeClass5xx = "5xx"
+)
+
+func main() {
+	ctx := context.Background()
+
+	client, err := monitoring.NewMetricClient(ctx)
+	if err != nil {
+		fmt.Printf("Unable to create NewMetricClient: %v\n", err)
+		os.Exit(1)
+	}
+	defer client.Close()
+
+	timeToEnd := time.Now().Add(timeToMonitor)
+	sampleIndex := 1
+	var timeWhenMaxErrorPercentageWasExceeded time.Time
+
+	for time.Now().Before(timeToEnd) {
+		endTime := time.Now()
+		// We look back the window amount.
+		startTime := endTime.Add(-samplingWindow)
+
+		res, err := metricsAreWithinThresholdForPeriod(ctx, client, sampleIndex, startTime, endTime)
```

Comment:
Perhaps this is an edge case, but I think this may be more intuitive if it were processed as rolling window. Ex:

with this setup with a sample size of 5 and a threshold of 50%

[0 0 0 1 1 | 1 1 0 0 0] passes as both samples are < 50%, but there is a sub-sequence  [1 1 1 1 0] in there that would have failed

Reply:
I have changed to have the sliding window be handled on the query client itself. Let me know if this is more intuitive